src/parameterized/uniform.jl

```diff
@@ -27,4 +27,43 @@ TV.as(::Uniform{()}) = as𝕀
 Base.rand(rng::AbstractRNG, T::Type, μ::Uniform{()}) = rand(rng, T)
 
 ###############################################################################
-# Uniform
+# Uniform over an interval
+#
+# We use an affine parameterization (shift, scale) versus a more traditional (lower, upper)
+# because of how `asparams` works. Since `asparams` works on a per-parameter basis,
+# we can enforce the `lower < upper` constraint by making `scale` be non-negative
+
+@kwstruct Uniform(shift, scale)
+
+function basemeasure(d::Uniform{(:shift, :scale)})
+    # Lebesgue measure on (shift, scale+shift)
+    lb = d.shift
+    ub = d.shift + d.scale
+    inbounds(x) = lb < x < ub
+    constℓ = 0.0
+    varℓ() = 0.0
+    base = Lebesgue(ℝ)
+    FactoredBase(inbounds, constℓ, varℓ, base)
+end
+
+asparams(::Type{<:Uniform}, ::Val{:shift}) = asℝ
+asparams(::Type{<:Uniform}, ::Val{:scale}) = asℝ₊
+
+"""
+The uniform probability measure on the interval (a, b).
+"""
+function Uniform(a, b)
+    @assert(a < b)
+    Uniform(shift=a, scale=b-a)
+end
+
+logdensity(d::Uniform{(:shift, :scale)}, x) = begin
+    @assert (d.scale > 0.0)
```

Comment:
Perhaps these should be checked in inner constructor to avoid the check on calls to `logdensity`.

I prefer invalid values such as `-Inf` or `NaN` over errors.
Density functions are used in many iterative algorithms that can on occasion produce invalid inputs.
`-Inf` or `NaN` will cause these algorithms to reject the offending parameter values and move on.
Errors will cause termination, or require expensive `try`/`catch` and then extra effort to avoid type instabilities.

Reply:
I haven't discussed this with @cscherrer , so I don't know his take, but it is my number 1 gripe with Distributions.jl.
I have to add code like `isnan(x) ? 0.0 #= results in -Inf =# :  abs(x)` for values that must be positive all over the place to work around it, or opt in via keyword arguments.
Neither of which is something I can rely on end users doing.
Meaning it may be worth eventually abandoning Distributions.jl, or trying to add passes that make the necessary adjustments.

All a bunch of extra work and complexity.

Reply:
Thank you both! Yes, I agree it's good to avoid try/catch. In this case, I think it's not really a problem to have a negative scale, since that's really just a linear coefficient in an affine transform. 

BTW, things in the works are in the `dev2` branch here and in MeasureBase. One thing that seems worth discussing here is the idea of a `proxy`.

For a while `Normal` was its own thing. Lots of parameterizations, defined independently. But that's kind of annoying, since scale/location families come up all over the place, and it's weird to implement them so many times.

That led to development of `Affine`. But then it's equally weird to ask for a `Normal(2,3)` and get an `Affine` as a result. Maybe we want to send the result into the function, so we'll need dispatch. It's good to avoid having to dig too deeply  into the type, when we have a chance.

So now (in `dev2`), `Normal(2,3)` builds a `Normal`, but `basemeasure` or `logdensity_def` just forward that `Normal` to its `proxy`, which is an `Affine`. Like this:
```julia
julia> d = Normal(2,3)
Normal(μ = 2, σ = 3)

julia> MeasureBase.proxy(d)
Affine((μ = 2, σ = 3), Normal{(), Tuple{}}(()))
```
[That should be `Affine((μ = 2, σ = 3), Normal())`, looks like a `show` bug]

Lately I've been working on composability and performance for exponential families, so some of this will change again. But that's not so relevant for `Uniform` unless the bounds are fixed.

Also, I should mention that new releases are AGPL, so I can offer dual licensing for commercial users. This is in an attempt to make it more likely I'll be able to keep doing this work. For that purpose I think it's the right call, but unfortunately it does mean I'll need to get something set up for people to contribute, in order to allow those contributions to be relicensed commercially. I'm still learning about all of this, but I'd be happy to discuss if you have any feedback or ideas.